src/intrusive.rs

```diff
@@ -36,27 +48,77 @@ struct Inner<T> {
 impl<T> List<T> {
     /// Create a new, empty event listener list.
     pub(super) fn new() -> Self {
-        Self(Mutex::new(Inner {
+        let inner = Inner {
             head: None,
             tail: None,
             next: None,
             len: 0,
             notified: 0,
-        }))
+        };
+
+        #[cfg(feature = "critical-section")]
+        let inner = RefCell::new(inner);
+
+        Self(Mutex::new(inner))
     }
 
     /// Get the total number of listeners without blocking.
+    #[cfg(all(feature = "std", not(feature = "critical-section")))]
     pub(crate) fn try_total_listeners(&self) -> Option<usize> {
         self.0.try_lock().ok().map(|list| list.len)
     }
 
+    /// Get the total number of listeners without blocking.
+    #[cfg(feature = "critical-section")]
+    pub(crate) fn try_total_listeners(&self) -> Option<usize> {
+        Some(critical_section::with(|cs| self.0.borrow(cs).borrow().len))
+    }
+
     /// Get the total number of listeners with blocking.
+    #[cfg(all(feature = "std", not(feature = "critical-section")))]
     pub(crate) fn total_listeners(&self) -> usize {
         self.0.lock().unwrap_or_else(|e| e.into_inner()).len
     }
+
+    /// Get the total number of listeners with blocking.
+    #[cfg(all(feature = "std", feature = "critical-section"))]
```

Comment:
This line looks wrong.
Also, I think it would be a good idea to define `total_listeners` explicitly, and `try_total_listeners` via `total_listeners` in the `critical-section` case.

Reply:
Good point, fixed.